### PR TITLE
MINOR: Cache topic resolution in TopicIds set

### PR DESCRIPTION
**Pull request by squah-confluent**
Looking up topics in a TopicsImage is relatively slow. Cache the results
in TopicIds to improve assignor performance. In benchmarks, we see a
noticeable improvement in performance in the heterogeneous case.

Before
```
Benchmark                                       (assignmentType)  (assignorType)  (isRackAware)  (memberCount)  (partitionsToMemberRatio)  (subscriptionType)  (topicCount)  Mode  Cnt    Score   Error  Units
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           RANGE          false          10000                         10         HOMOGENEOUS          1000  avgt    5   36.400 ± 3.004  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           RANGE          false          10000                         10       HETEROGENEOUS          1000  avgt    5  158.340 ± 0.825  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL         UNIFORM          false          10000                         10         HOMOGENEOUS          1000  avgt    5    1.329 ± 0.041  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL         UNIFORM          false          10000                         10       HETEROGENEOUS          1000  avgt    5  382.901 ± 6.203  ms/op
```

After
```
Benchmark                                       (assignmentType)  (assignorType)  (isRackAware)  (memberCount)  (partitionsToMemberRatio)  (subscriptionType)  (topicCount)  Mode  Cnt    Score   Error  Units
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           RANGE          false          10000                         10         HOMOGENEOUS          1000  avgt    5   36.465 ± 1.954  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           RANGE          false          10000                         10       HETEROGENEOUS          1000  avgt    5  114.043 ± 1.424  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL         UNIFORM          false          10000                         10         HOMOGENEOUS          1000  avgt    5    1.454 ± 0.019  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL         UNIFORM          false          10000                         10       HETEROGENEOUS          1000  avgt    5  342.840 ± 2.744  ms/op
```

---

Based heavily on https://github.com/apache/kafka/pull/16527.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java [hunk 1]
```diff
@@ -21,25 +21,169 @@
...
+         */
+        void clear();
+    }
+
+    /**
+     * A base implementation of TopicResolver.
+     *
+     * Provides an implementation of equals and hashCode based on the underlying TopicsImage.
+     */
+    public abstract static class BaseTopicResolver implements TopicResolver {
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java]: nit: Should we keep it private?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java [hunk 2]
```diff
@@ -164,20 +308,20 @@ public boolean equals(Object o) {
...
-        result = 31 * result + image.hashCode();
+        result = 31 * result + resolver.hashCode();
         return result;
     }
 
     @Override
     public String toString() {
         return "TopicIds(topicNames=" + topicNames +
-            ", image=" + image +
+            ", resolver=" + resolver +
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java:316]: nit: If we toString it here, should we add toString methods to the resolvers too?

**Review by dajac — COMMENTED** [review]
@squah-confluent Thanks for the patch. I left a few nits.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java (hunk 1 above)
Comment (squah-confluent) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java]: Yes, it doesn't need to be public.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java (hunk 2 above)
Comment (squah-confluent) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java:316]: I'll add a toString implementation that returns `<class name>(image=<image>)`.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java [hunk 3]
```diff
@@ -21,25 +21,175 @@
...
+        void clear();
+    }
+
+    /**
+     * A base implementation of TopicResolver.
+     *
+     * Provides an implementation of equals, hashCode and toString based on the underlying
+     * TopicsImage.
+     */
+    private abstract static class BaseTopicResolver implements TopicResolver {
```
Comment (mumrah) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java]: Why do we need this abstract class? Can we just have the interface and concrete class?

**Review by mumrah — COMMENTED** [review]
@squah-confluent, Seeing that the cache is not actually used outside of tests and benchmarks, I'm guessing this is still WIP.

Does this come down to performance differences between HashMap and PCollectionsImmutableMap?

If we decide we really need faster topic ID to name lookups, I would consider adding it to TopicsImage. Managing a cache outside of the image will be a bit difficult. 

Another thing to consider is the lifetime of the cache. Do we really need the ID + name mappings kept in memory forever?

**Comment by dajac** [comment]
@mumrah 

> Another thing to consider is the lifetime of the cache. Do we really need the ID + name mappings kept in memory forever?

The lifetime of the cache is bound to the call. It is not kept forever.

> Does this come down to performance differences between HashMap and PCollectionsImmutableMap?

Yes.

> If we decide we really need faster topic ID to name lookups, I would consider adding it to TopicsImage. Managing a cache outside of the image will be a bit difficult.

We could consider this separately. At the moment, we don't really have the time to do it. The current strategy seems to be a good tradeoff at the moment given that it is only bound to the call and not kept forever.

**Comment by mumrah** [comment]
@dajac thanks for the explanation, makes sense. Can we include a javadoc on the class describing the expected lifetime of this class?

**Comment by squah-confluent** [comment]
@mumrah 
> Seeing that the cache is not actually used outside of tests and benchmarks, I'm guessing this is still WIP.

It's used in TargetAssignmentBuilder.build(), which is used by the new group coordinator.

I've updated the javadoc to describe the lifetime of the cache.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java (hunk 3 above)
Comment (squah-confluent) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java]: I can remove it.

**Review by dajac — APPROVED** [review]
LGTM, thanks.

**Comment by dajac** [comment]
@mumrah I will merge it. If you have further comments, @squah-confluent can address separately.

**Pull request merged**